File: luajit_programs_and_containers/building_blocks/s_expression/lua_runtime/se_builtins_dict.lua
-- ============================================================================
-- se_builtins_dict.lua
-- Mirrors s_engine_builtins_dict.h / se_load_dictionary.c
--
-- Dictionary loading and value extraction oneshot functions.
-- All functions: fn(inst, node)  (oneshot, no return value)
--
-- In C, dictionaries are ROM binary structures (s_expr_param_t arrays).
-- In Lua, dictionaries are plain Lua tables stored in the blackboard:
--
--   {                              -- top-level dict
--     ["key_name"] = 42,           -- scalar value
--     [0xABCD1234] = 3.14,         -- hash-keyed scalar
--     ["sub"] = { ... },           -- nested dict
--   }
--
-- se_load_dictionary stores a reference (the Lua table itself) into a
-- blackboard field. se_dict_extract_* navigate the table and write values.
--
-- String-path functions (params[2] = str_idx):
--   Navigate using dot-separated keys  "level1.level2.key"
--
-- Hash-path functions (params[2..N-1] = str_hash):
--   Navigate using a sequence of hash keys
-- ============================================================================

local se_runtime = require("se_runtime")

local param_str        = se_runtime.param_str
local param_field_name = se_runtime.param_field_name
local field_get        = se_runtime.field_get

local M = {}

-- ============================================================================
-- Internal helpers
-- ============================================================================

-- Retrieve the dict table from a blackboard PTR field
local function get_dict_from_bb(inst, node, param_idx)
    local field_name = param_field_name(node, param_idx)
    local dict = inst.blackboard[field_name]
    assert(dict and type(dict) == "table",
        "se_dict: blackboard field is not a table: " .. tostring(field_name))
    return dict
end

-- Navigate a nested dict using a dot-separated string path
-- "a.b.c" -> dict["a"]["b"]["c"]
local function navigate_string_path(dict, path)
    local cur = dict
    for key in path:gmatch("[^%.]+") do
        if type(cur) ~= "table" then return nil end
        -- Try string key first, then numeric key
        cur = cur[key]
        if cur == nil then cur = cur and cur[tonumber(key)] end
    end
    return cur
end

-- Navigate a nested dict using an array of hash keys
local function navigate_hash_path(dict, hashes)
    local cur = dict
    for _, h in ipairs(hashes) do
        if type(cur) ~= "table" then return nil end
        cur = cur[h]
        if cur == nil then cur = cur and cur[tostring(h)] end
    end
    return cur
end

-- Extract scalar from dict entry; entry may be raw value or {value=V} table
local function extract_value(entry)
    if type(entry) == "table" and entry.value ~= nil then
        return entry.value
    end
    return entry
end

-- Collect hash key params from node.params[start_idx .. end_idx]
-- Returns array of hash numbers
local function collect_hashes(node, start_idx, end_idx)
    local hashes = {}
    for i = start_idx, end_idx do
        local p = (node.params or {})[i]
        if not p then break end
        hashes[#hashes + 1] = (type(p.value) == "table") and p.value.hash or p.value
    end
    return hashes
end

-- Find the index of the last field_ref param (used by hash-path functions)
local function last_field_param_idx(node)
    local params = node.params or {}
    for i = #params, 1, -1 do
        if params[i].type == "field_ref" or params[i].type == "nested_field_ref" then
            return i
        end
    end
    return nil
end

-- ============================================================================
-- SE_LOAD_DICTIONARY  (oneshot)
-- Stores a reference to a dictionary constant into a blackboard PTR field.
-- params[1] = field_ref  (destination PTR field)
-- params[2] = const_ref | str_idx | str_hash  (name/hash of the constant dict)
--
-- In C this stores a pointer into the binary module image.
-- In Lua we look up mod.module_data.constants and store the Lua table.
-- If params[2] is a list_start, use its items table directly (inline dict).
-- ============================================================================
M.se_load_dictionary = function(inst, node)
    assert(#(node.params or {}) >= 2,
        "se_load_dictionary: requires [field_ref dest] [const ref]")

    local field_name = param_field_name(node, 1)
    local p2 = node.params[2]
    local dict_ref

    local consts = inst.mod.module_data and inst.mod.module_data.constants

    if p2.type == "str_idx" or p2.type == "str_ptr" then
        -- String constant name: direct lookup
        dict_ref = consts and consts[p2.value]
    elseif p2.type == "const_ref" then
        dict_ref = consts and consts[p2.value]
    elseif p2.type == "str_hash" or p2.type == "uint" or p2.type == "int" then
        -- Hash: search constants table for name_hash match, or use as direct key
        local hash = (type(p2.value) == "table") and p2.value.hash or p2.value
        if consts then
            dict_ref = consts[hash]
            if not dict_ref then
                for _, v in pairs(consts) do
                    if type(v) == "table" and v.name_hash == hash then
                        dict_ref = v.data or v; break
                    end
                end
            end
        end
    elseif p2.type == "list_start" then
        -- Inline dict embedded directly in params
        dict_ref = p2.items or {}
    end

    assert(dict_ref, "se_load_dictionary: could not resolve dictionary constant")
    inst.blackboard[field_name] = dict_ref
end

-- ============================================================================
-- String-path extraction
-- params[1] = field_ref  (source dict in blackboard)
-- params[2] = str_idx    (dot-separated path string)
-- params[3] = field_ref  (destination field)
-- ============================================================================

M.se_dict_extract_int = function(inst, node)
    local dict  = get_dict_from_bb(inst, node, 1)
    local path  = param_str(node, 2)
    local entry = navigate_string_path(dict, path)
    local val   = entry and math.floor(tonumber(extract_value(entry)) or 0) or 0
    inst.blackboard[param_field_name(node, 3)] = val
end

M.se_dict_extract_uint = function(inst, node)
    local dict  = get_dict_from_bb(inst, node, 1)
    local path  = param_str(node, 2)
    local entry = navigate_string_path(dict, path)
    local val   = entry and math.floor(math.abs(tonumber(extract_value(entry)) or 0)) or 0
    inst.blackboard[param_field_name(node, 3)] = val
end

M.se_dict_extract_float = function(inst, node)
    local dict  = get_dict_from_bb(inst, node, 1)
    local path  = param_str(node, 2)
    local entry = navigate_string_path(dict, path)
    local val   = entry and ((tonumber(extract_value(entry)) or 0) + 0.0) or 0.0
    inst.blackboard[param_field_name(node, 3)] = val
end

M.se_dict_extract_bool = function(inst, node)
    local dict  = get_dict_from_bb(inst, node, 1)
    local path  = param_str(node, 2)
    local entry = navigate_string_path(dict, path)
    local raw   = entry and extract_value(entry) or nil
    local val   = (raw and raw ~= 0 and raw ~= false) and 1 or 0
    inst.blackboard[param_field_name(node, 3)] = val
end

M.se_dict_extract_hash = function(inst, node)
    local dict  = get_dict_from_bb(inst, node, 1)
    local path  = param_str(node, 2)
    local entry = navigate_string_path(dict, path)
    inst.blackboard[param_field_name(node, 3)] = entry and extract_value(entry) or 0
end

-- ============================================================================
-- Hash-path extraction
-- params[1]       = field_ref  (source dict)
-- params[2..N-1]  = str_hash   (path segment hashes)
-- params[N]       = field_ref  (destination field)  -- last field_ref param
-- ============================================================================

local function hash_path_extract(inst, node, converter)
    local dict     = get_dict_from_bb(inst, node, 1)
    local dest_idx = last_field_param_idx(node)
    -- dest_idx must be > 1 (at least one hash param between them)
    assert(dest_idx and dest_idx > 2,
        "se_dict_extract_h: missing hash path params or destination")
    local hashes = collect_hashes(node, 2, dest_idx - 1)
    local entry  = navigate_hash_path(dict, hashes)
    inst.blackboard[param_field_name(node, dest_idx)] =
        converter(entry and extract_value(entry) or nil)
end

M.se_dict_extract_int_h = function(inst, node)
    hash_path_extract(inst, node,
        function(v) return v and math.floor(tonumber(v) or 0) or 0 end)
end

M.se_dict_extract_uint_h = function(inst, node)
    hash_path_extract(inst, node,
        function(v) return v and math.floor(math.abs(tonumber(v) or 0)) or 0 end)
end

M.se_dict_extract_float_h = function(inst, node)
    hash_path_extract(inst, node,
        function(v) return v and ((tonumber(v) or 0) + 0.0) or 0.0 end)
end

M.se_dict_extract_bool_h = function(inst, node)
    hash_path_extract(inst, node,
        function(v) return (v and v ~= 0 and v ~= false) and 1 or 0 end)
end

M.se_dict_extract_hash_h = function(inst, node)
    hash_path_extract(inst, node, function(v) return v or 0 end)
end

-- ============================================================================
-- SE_DICT_STORE_PTR  (oneshot)
-- Stores a reference to a sub-table at a string path into a blackboard field.
-- params[1] = field_ref  (source dict)
-- params[2] = str_idx    (dot-separated path)
-- params[3] = field_ref  (destination PTR field)
-- ============================================================================
M.se_dict_store_ptr = function(inst, node)
    local dict = get_dict_from_bb(inst, node, 1)
    local path = param_str(node, 2)
    local sub  = navigate_string_path(dict, path)
    inst.blackboard[param_field_name(node, 3)] = sub   -- nil if not found
end

-- ============================================================================
-- SE_DICT_STORE_PTR_H  (oneshot)
-- Stores a sub-table reference navigated by hash path.
-- params[1]       = field_ref  (source dict)
-- params[2..N-1]  = str_hash   (path segments)
-- params[N]       = field_ref  (destination PTR field)
-- ============================================================================
M.se_dict_store_ptr_h = function(inst, node)
    local dict     = get_dict_from_bb(inst, node, 1)
    local dest_idx = last_field_param_idx(node)
    assert(dest_idx and dest_idx > 2,
        "se_dict_store_ptr_h: missing hash path params or destination")
    local hashes = collect_hashes(node, 2, dest_idx - 1)
    local sub    = navigate_hash_path(dict, hashes)
    inst.blackboard[param_field_name(node, dest_idx)] = sub
end

return M

File: micropython/building_blocks/dsl_engines/s_engine/s_expression_micropython/mpy_runtime/se_builtins_dict.py
# ==========================================================================
# se_builtins_dict.py
# Mirrors se_load_dictionary.c / s_engine_builtins_dict.h
# ==========================================================================

import se_runtime
import math

_N_PARAMS = se_runtime.N_PARAMS
_P_TYPE = se_runtime.P_TYPE
_P_VALUE = se_runtime.P_VALUE


def _s_expr_hash(s):
    """FNV-1a 32-bit hash."""
    h = 2166136261
    for c in s:
        h ^= ord(c)
        h = ((h << 24) + h * 403) & 0xFFFFFFFF
    return h


def _parse_scalar(p):
    t = p[_P_TYPE]
    if t == "int" or t == "uint" or t == "float":
        return p[_P_VALUE]
    if t == "str_idx" or t == "str_ptr":
        s = p[_P_VALUE]
        return {"str": s, "hash": _s_expr_hash(s)}
    return None


def _parse_dict(params, start_i):
    result = {}
    i = start_i + 1
    while i < len(params):
        p = params[i]
        t = p[_P_TYPE]
        if t == "dict_end":
            return result, i + 1
        if t == "dict_key" or t == "dict_key_hash":
            key = p[_P_VALUE]
            i += 1
            if i >= len(params):
                break
            vp = params[i]
            if vp[_P_TYPE] == "dict_start":
                val, i = _parse_dict(params, i)
            elif vp[_P_TYPE] == "array_start":
                val, i = _parse_array(params, i)
            else:
                val = _parse_scalar(vp)
                i += 1
            if i < len(params) and params[i][_P_TYPE] == "end_dict_key":
                i += 1
            result[key] = val
        else:
            i += 1
    return result, i


def _parse_array(params, start_i):
    result = {}
    idx = 0
    i = start_i + 1
    while i < len(params):
        p = params[i]
        t = p[_P_TYPE]
        if t == "array_end":
            return result, i + 1
        if t == "dict_start":
            val, i = _parse_dict(params, i)
        elif t == "array_start":
            val, i = _parse_array(params, i)
        else:
            val = _parse_scalar(p)
            i += 1
        if val is not None:
            result[idx] = val
            idx += 1
    return result, i


def _as_number(v):
    if isinstance(v, dict) and "str" in v:
        try:
            return float(v["str"])
        except (ValueError, TypeError):
            return 0
    try:
        return float(v) if v is not None else 0
    except (ValueError, TypeError):
        return 0


def _as_hash(v):
    if isinstance(v, dict) and "hash" in v:
        return v["hash"]
    if isinstance(v, str):
        return _s_expr_hash(v)
    return int(float(v)) if v is not None else 0


def _navigate_str_path(d, path):
    cur = d
    for key in path.split("."):
        if not isinstance(cur, dict):
            return None
        v = cur.get(key)
        if v is None:
            try:
                v = cur.get(int(key))
            except (ValueError, TypeError):
                pass
        cur = v
    return cur


def _collect_path_items(node, start_idx, end_idx):
    items = []
    params = node[_N_PARAMS]
    for i in range(start_idx, end_idx + 1):
        if i >= len(params):
            break
        p = params[i]
        v = p[_P_VALUE]
        if isinstance(v, tuple) and len(v) == 2:
            items.append({"hash": v[0], "str": v[1]})
        elif isinstance(v, dict):
            items.append({"hash": v.get("hash", 0), "str": v.get("str")})
        else:
            items.append({"hash": v, "str": None})
    return items


def _navigate_hash_path(d, path_items):
    cur = d
    for item in path_items:
        if not isinstance(cur, dict):
            return None
        v = cur.get(item["hash"])
        if v is None and item["str"]:
            v = cur.get(item["str"])
            if v is None:
                try:
                    v = cur.get(int(item["str"]))
                except (ValueError, TypeError):
                    pass
        cur = v
    return cur


def _last_field_idx(node):
    params = node[_N_PARAMS]
    for i in range(len(params) - 1, -1, -1):
        t = params[i][_P_TYPE]
        if t == "field_ref" or t == "nested_field_ref":
            return i + 1  # 1-based
    return None


def _bb_dict(inst, node, param_idx):
    fname = se_runtime.param_field_name(node, param_idx)
    d = inst["blackboard"].get(fname)
    assert isinstance(d, dict), "se_dict: field '%s' is not dict" % str(fname)
    return d


def se_load_dictionary(inst, node):
    params = node[_N_PARAMS]
    assert len(params) >= 2
    fname = se_runtime.param_field_name(node, 1)
    p2 = params[1]
    assert p2[_P_TYPE] == "dict_start"
    d, _ = _parse_dict(params, 1)
    inst["blackboard"][fname] = d

se_load_dictionary_hash = se_load_dictionary


def se_dict_extract_int(inst, node):
    d = _bb_dict(inst, node, 1)
    v = _navigate_str_path(d, se_runtime.param_str(node, 2))
    inst["blackboard"][se_runtime.param_field_name(node, 3)] = int(_as_number(v)) if v is not None else 0

def se_dict_extract_uint(inst, node):
    d = _bb_dict(inst, node, 1)
    v = _navigate_str_path(d, se_runtime.param_str(node, 2))
    inst["blackboard"][se_runtime.param_field_name(node, 3)] = int(abs(_as_number(v))) if v is not None else 0

def se_dict_extract_float(inst, node):
    d = _bb_dict(inst, node, 1)
    v = _navigate_str_path(d, se_runtime.param_str(node, 2))
    inst["blackboard"][se_runtime.param_field_name(node, 3)] = float(_as_number(v)) if v is not None else 0.0

def se_dict_extract_bool(inst, node):
    d = _bb_dict(inst, node, 1)
    v = _navigate_str_path(d, se_runtime.param_str(node, 2))
    n = _as_number(v) if v is not None else 0
    inst["blackboard"][se_runtime.param_field_name(node, 3)] = 1 if n != 0 else 0

def se_dict_extract_hash(inst, node):
    d = _bb_dict(inst, node, 1)
    v = _navigate_str_path(d, se_runtime.param_str(node, 2))
    inst["blackboard"][se_runtime.param_field_name(node, 3)] = _as_hash(v) if v is not None else 0


def _hash_extract(inst, node, conv):
    d = _bb_dict(inst, node, 1)
    dest = _last_field_idx(node)
    assert dest and dest > 2
    path = _collect_path_items(node, 1, dest - 2)
    v = _navigate_hash_path(d, path)
    inst["blackboard"][se_runtime.param_field_name(node, dest)] = conv(v)

def se_dict_extract_int_h(inst, node):
    _hash_extract(inst, node, lambda v: int(_as_number(v)) if v is not None else 0)
def se_dict_extract_uint_h(inst, node):
    _hash_extract(inst, node, lambda v: int(abs(_as_number(v))) if v is not None else 0)
def se_dict_extract_float_h(inst, node):
    _hash_extract(inst, node, lambda v: float(_as_number(v)) if v is not None else 0.0)
def se_dict_extract_bool_h(inst, node):
    _hash_extract(inst, node, lambda v: 1 if (_as_number(v) if v is not None else 0) != 0 else 0)
def se_dict_extract_hash_h(inst, node):
    _hash_extract(inst, node, lambda v: _as_hash(v) if v is not None else 0)

def se_dict_store_ptr(inst, node):
    d = _bb_dict(inst, node, 1)
    sub = _navigate_str_path(d, se_runtime.param_str(node, 2))
    inst["blackboard"][se_runtime.param_field_name(node, 3)] = sub if isinstance(sub, dict) else None

def se_dict_store_ptr_h(inst, node):
    d = _bb_dict(inst, node, 1)
    dest = _last_field_idx(node)
    assert dest and dest > 2
    path = _collect_path_items(node, 1, dest - 2)
    sub = _navigate_hash_path(d, path)
    inst["blackboard"][se_runtime.param_field_name(node, dest)] = sub if isinstance(sub, dict) else None


def se_load_function_dict(inst, node):
    params = node[_N_PARAMS]
    children = node[se_runtime.N_CHILDREN]
    assert len(params) >= 2, "se_load_function_dict: need >= 2 params"

    fname = se_runtime.param_field_name(node, 1)

    # Collect dict_key names in order; each pairs with the next child
    keys = []
    for p in params:
        if p[_P_TYPE] == "dict_key":
            keys.append(p[_P_VALUE])

    assert len(keys) == len(children), \
        "se_load_function_dict: %d keys but %d children" % (len(keys), len(children))

    d = {}
    for i in range(len(keys)):
        key_hash = _s_expr_hash(keys[i])
        child_node = children[i]
        def _make_fn(cn):
            def fn(calling_inst, exec_node, eid, edata):
                return se_runtime.invoke_any(calling_inst, cn, eid, edata)
            return fn
        d[key_hash] = _make_fn(child_node)

    inst["blackboard"][fname] = d


builtins = {
    "se_load_dictionary": se_load_dictionary,
    "se_load_dictionary_hash": se_load_dictionary_hash,
    "se_load_function_dict": se_load_function_dict,
    "se_dict_extract_int": se_dict_extract_int,
    "se_dict_extract_uint": se_dict_extract_uint,
    "se_dict_extract_float": se_dict_extract_float,
    "se_dict_extract_bool": se_dict_extract_bool,
    "se_dict_extract_hash": se_dict_extract_hash,
    "se_dict_extract_int_h": se_dict_extract_int_h,
    "se_dict_extract_uint_h": se_dict_extract_uint_h,
    "se_dict_extract_float_h": se_dict_extract_float_h,
    "se_dict_extract_bool_h": se_dict_extract_bool_h,
    "se_dict_extract_hash_h": se_dict_extract_hash_h,
    "se_dict_store_ptr": se_dict_store_ptr,
    "se_dict_store_ptr_h": se_dict_store_ptr_h,
}
